src/Layers/xrRender/ParticleEffect.cpp

```diff
@@ -453,126 +453,136 @@ __forceinline void magnitude_sse(Fvector& vec, float& res)
+                M.j.normalize();
+                if (pPE->m_RT_Flags.is(CParticleEffect::flRT_XFORM))
+                {
+                    Fvector p;
+                    pPE->m_XFORM.transform_tiny(p, m.pos);
+                    M.mulA_43(pPE->m_XFORM);
+                    FillSprite(pv, M.j, M.i, p, lt, rb, r_x, r_y, m.color, sina, cosa);
                 }
                 else
                 {
-                    Fvector dir;
-                    if (speed >= EPS_S)
-                        dir.div(m.vel, speed);
-                    else
-                        dir.setHP(-pPE->m_Def->m_APDefaultRotation.y, -pPE->m_Def->m_APDefaultRotation.x);
-                    if (pPE->m_RT_Flags.is(CParticleEffect::flRT_XFORM))
-                    {
-                        Fvector p, d;
-                        pPE->m_XFORM.transform_tiny(p, m.pos);
-                        pPE->m_XFORM.transform_dir(d, dir);
-                        FillSprite(pv, p, d, lt, rb, r_x, r_y, m.color, sina, cosa);
-                    }
-                    else
-                    {
-                        FillSprite(pv, m.pos, dir, lt, rb, r_x, r_y, m.color, sina, cosa);
-                    }
+                    FillSprite(pv, M.j, M.i, m.pos, lt, rb, r_x, r_y, m.color, sina, cosa);
                 }
             }
             else
             {
+                Fvector dir;
+                if (speed >= EPS_S)
+                    dir.div(m.vel, speed);
+                else
+                    dir.setHP(-pPE->m_Def->m_APDefaultRotation.y, -pPE->m_Def->m_APDefaultRotation.x);
                 if (pPE->m_RT_Flags.is(CParticleEffect::flRT_XFORM))
                 {
-                    Fvector p;
+                    Fvector p, d;
                     pPE->m_XFORM.transform_tiny(p, m.pos);
-                    FillSprite(pv, RDEVICE.vCameraTop, RDEVICE.vCameraRight, p, lt, rb, r_x, r_y, m.color, sina, cosa);
+                    pPE->m_XFORM.transform_dir(d, dir);
+                    FillSprite(pv, p, d, lt, rb, r_x, r_y, m.color, sina, cosa);
                 }
                 else
                 {
-                    FillSprite(pv, RDEVICE.vCameraTop, RDEVICE.vCameraRight, m.pos, lt, rb, r_x, r_y, m.color, sina, cosa);
+                    FillSprite(pv, m.pos, dir, lt, rb, r_x, r_y, m.color, sina, cosa);
                 }
             }
         }
-    FOR_END
+        else
+        {
+            if (pPE->m_RT_Flags.is(CParticleEffect::flRT_XFORM))
+            {
+                Fvector p;
+                pPE->m_XFORM.transform_tiny(p, m.pos);
+                FillSprite(pv, RDEVICE.vCameraTop, RDEVICE.vCameraRight, p, lt, rb, r_x, r_y, m.color, sina, cosa);
+            }
+            else
+            {
+                FillSprite(pv, RDEVICE.vCameraTop, RDEVICE.vCameraRight, m.pos, lt, rb, r_x, r_y, m.color, sina, cosa);
+            }
+        }
+    };
+
+    constexpr u32 max_count = 75;
+
+    if (true && count <= max_count)
```

Comment:
Strange, but this cycle is always faster in multithreaded mode.

Since, this is a WIP pull request, the code was pushed as is.

Reply:
Why you add 'true' to the condition?

Reply:
> Strange, but this cycle is always faster in multithreaded mode.
Since, this is a WIP pull request, the code was pushed as is.